flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/embedded/AbstractEmbeddedDataStreamPythonFunctionOperator.java

```diff
@@ -104,4 +134,45 @@ public Map<String, String> getJobParameters() {
         }
         return jobParameters;
     }
+
+    private class SideOutputContext {
+
+        private Map<String, byte[]> sideOutputTypeInfoPayloads = new HashMap<>();
+
+        private Map<String, PythonTypeUtils.DataConverter<Object, Object>>
+                sideOutputDataConverters = new HashMap<>();
+
+        @SuppressWarnings("unchecked")
+        public void open() {
+            for (Map.Entry<String, OutputTag<?>> entry : sideOutputTags.entrySet()) {
+                sideOutputTypeInfoPayloads.put(
+                        entry.getKey(), getSideOutputTypeInfoPayload(entry.getValue()));
+
+                sideOutputDataConverters.put(
+                        entry.getKey(),
+                        PythonTypeUtils.TypeInfoToDataConverter.typeInfoDataConverter(
+                                (TypeInformation) entry.getValue().getTypeInfo()));
+            }
+        }
+
+        @SuppressWarnings("unchecked")
+        public void collectSideOutputById(String id, Object record) {
+            OutputTag<?> outputTag = sideOutputTags.get(id);
```

Comment:
If reached (when KeyError mentioned above is solved), undefined output tag records will flow into the main stream.